File: generate_tts.js
// generate_tts.js
// Node script to generate per-word MP3s for words listed in public/words.txt
// Uses google-tts-api to obtain mp3 URLs from Google Translate TTS and downloads them.
// Usage:
//   npm install google-tts-api node-fetch@2 mkdirp
//   node generate_tts.js

const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');
const googleTTS = require('google-tts-api');
const mkdirp = require('mkdirp');

const BASE = process.cwd();
const WORDS_PATH = path.join(BASE, 'public', 'words.txt');
const OUT_DIR = path.join(BASE, 'public', 'tts');

if (!fs.existsSync(WORDS_PATH)) {
  console.error('Could not find public/words.txt. Run this from project root.');
  process.exit(1);
}

mkdirp.sync(OUT_DIR);

const words = fs.readFileSync(WORDS_PATH, 'utf8').split(/\r?\n/).map(s=>s.trim()).filter(Boolean);
if (!words.length) {
  console.error('No words found in public/words.txt');
  process.exit(1);
}

async function download(url, dest) {
  const res = await fetch(url, { headers: { 'User-Agent': 'node.js' } });
  if (!res.ok) throw new Error(`Failed to download ${url}: ${res.status}`);
  const buffer = await res.buffer();
  fs.writeFileSync(dest, buffer);
}

(async () => {
  console.log(`Generating ${words.length} mp3 files into ${OUT_DIR} ...`);
  for (let i = 0; i < words.length; i++) {
    const w = words[i];
    const fname = encodeURIComponent(w) + '.mp3';
    const outPath = path.join(OUT_DIR, fname);
    if (fs.existsSync(outPath)) {
      console.log('Skipping (exists):', outPath);
      continue;
    }
    try {
      console.log(`Generating (${i+1}/${words.length}):`, w);
      const url = googleTTS.getAudioUrl(w, { lang: 'he', slow: false, host: 'https://translate.google.com' });
      await download(url, outPath);
      // small delay to be nice to the service
      await new Promise(r=>setTimeout(r, 250));
    } catch (e) {
      console.error('Failed for', w, e.message || e);
    }
  }
  console.log('Done.');
})();

File: generate_tts.py
#!/usr/bin/env python3
"""
generate_tts.py

Generates per-word MP3 files for the words listed in public/words.txt using gTTS.
Files are written to public/tts/<encoded-word>.mp3 where encoding is urllib.parse.quote(...)

Usage:
  python generate_tts.py

Dependencies:
  pip install gTTS

"""
import os
from gtts import gTTS
from urllib.parse import quote

BASE = os.path.dirname(os.path.abspath(__file__))
WORDS_PATH = os.path.join(BASE, 'public', 'words.txt')
OUT_DIR = os.path.join(BASE, 'public', 'tts')

if not os.path.exists(WORDS_PATH):
    print('Could not find public/words.txt. Make sure you run this from the project root.')
    raise SystemExit(1)

os.makedirs(OUT_DIR, exist_ok=True)

with open(WORDS_PATH, 'r', encoding='utf-8') as f:
    words = [line.strip() for line in f if line.strip()]

if not words:
    print('No words found in public/words.txt')
    raise SystemExit(1)

print(f'Generating {len(words)} mp3 files into {OUT_DIR} ...')
for w in words:
    # Create a safe file name using percent-encoding
    fname = quote(w, safe='') + '.mp3'
    out_path = os.path.join(OUT_DIR, fname)
    if os.path.exists(out_path):
        print('Skipping (exists):', out_path)
        continue
    try:
        print('Generating:', w, '->', out_path)
        tts = gTTS(w, lang='he')
        tts.save(out_path)
    except Exception as e:
        print('Failed for', w, e)

print('Done.')
